source/opt/instruction.cpp

```diff
@@ -109,5 +110,166 @@ void Instruction::ReplaceOperands(const std::vector<Operand>& new_operands) {
+        break;
+      default:
+        done = true;
+        break;
+    }
+  }
+  return base_inst;
+}
+
+bool Instruction::IsReadOnlyVariable(IRContext* ctx) const {
+  assert(ctx->module()->HasCapability(SpvCapabilityShader) &&
+         "We don't know how enough about non-shaders to answer this question.");
+  uint32_t storage_class = GetSingleWordInOperand(0);
+  ir::Instruction* type_def = ctx->get_def_use_mgr()->GetDef(type_id());
+
+  switch (storage_class) {
+    case SpvStorageClassUniformConstant:
+      if (!type_def->IsVulkanStorageImage(ctx) &&
+          !type_def->IsVulkanStorageTexelBuffer(ctx)) {
+        return true;
+      }
+      break;
+    case SpvStorageClassUniform:
+      if (!type_def->IsVulkanStorageBuffer(ctx)) {
+        return true;
+      }
+      break;
+    case SpvStorageClassPushConstant:
+    case SpvStorageClassInput:
+      return true;
+    default:
+      break;
+  }
+
+  auto decorations = ctx->get_decoration_mgr()->GetDecorationsFor(
+      result_id(), SpvDecorationNonWritable);
+  return !decorations.empty();
+}
+
+bool Instruction::IsVulkanStorageImage(IRContext* ctx) const {
+  if (opcode() == SpvOpTypePointer) {
+    ir::Instruction* base_type =
+        ctx->get_def_use_mgr()->GetDef(GetSingleWordInOperand(1));
+    return base_type->IsVulkanStorageImage(ctx);
+  }
+
+  if (opcode() != SpvOpTypeImage) {
+    return false;
+  }
+
+  if (GetSingleWordInOperand(0) == SpvDimBuffer) {
+    return false;
+  }
+
+  // Check if the image is sampled.  If we do not know for sure that it is,
+  // then assume it is a storage image.
+  return GetSingleWordInOperand(6) != 1;
+}
+
+bool Instruction::IsVulkanSampledImage(IRContext* ctx) const {
+  if (opcode() == SpvOpTypePointer) {
+    ir::Instruction* base_type =
+        ctx->get_def_use_mgr()->GetDef(GetSingleWordInOperand(1));
+    return base_type->IsVulkanSampledImage(ctx);
+  }
+
+  if (opcode() != SpvOpTypeImage) {
+    return false;
+  }
+
+  if (GetSingleWordInOperand(0) == SpvDimBuffer) {
+    return false;
+  }
+
+  // Check if the image is sampled.  If we know for sure that it is,
+  // then return true.
+  return GetSingleWordInOperand(6) == 1;
+}
+
+bool Instruction::IsVulkanStorageTexelBuffer(IRContext* ctx) const {
+  if (opcode() == SpvOpTypePointer) {
+    ir::Instruction* base_type =
+        ctx->get_def_use_mgr()->GetDef(GetSingleWordInOperand(1));
+    return base_type->IsVulkanStorageTexelBuffer(ctx);
+  }
+
+  if (opcode() != SpvOpTypeImage) {
+    return false;
+  }
+
+  if (GetSingleWordInOperand(0) != SpvDimBuffer) {
+    return false;
+  }
+
+  // Check if the image is sampled.  If we do not know for sure that it is,
+  // then assume it is a storage texel buffer.
+  return GetSingleWordInOperand(6) != 1;
+}
+
+bool Instruction::IsVulkanStorageBuffer(IRContext* ctx) const {
+  // Is there a difference between a "Storage buffer" and a "dynamic storage
+  // buffer" in SPIR-V and do we age about the difference?
```

Comment:
"do we age"?

Reply:
Fixed, but I'm still looking for an answer to the question.